File: server/src/services/assessment/utils/crud/release_results.rs
use uuid::Uuid;
use crate::utils::error::{AppError, AppResult};
use crate::utils::fmt_utc;
use crate::schema::assessment_schema::*;

impl crate::services::assessment::AssessmentService {
    pub async fn release_results(
        &self,
        assessment_id: Uuid,
        teacher_id: Uuid,
    ) -> AppResult<AssessmentResponse> {
        let assessment = self.assessment_repo.find_by_id(assessment_id).await?
            .ok_or_else(|| AppError::NotFound("Assessment not found".to_string()))?;

        let _class = self.class_repo.find_by_id(assessment.class_id).await?
            .ok_or_else(|| AppError::NotFound("Class not found".to_string()))?;

        if !self.class_repo.is_teacher_of_class(teacher_id, assessment.class_id).await? {
            return Err(AppError::Forbidden("Access denied".to_string()));
        }

        if !assessment.is_published {
            return Err(AppError::BadRequest("Assessment must be published first".to_string()));
        }

        let released = self.assessment_repo.release_results(assessment_id).await?;

        let question_count = self.assessment_repo
            .find_questions_by_assessment_id(assessment_id).await?.len();
        let submission_count = self.assessment_repo
            .count_submissions_by_assessment_id(assessment_id).await?;

        Ok(AssessmentResponse {
            id: released.id,
            class_id: released.class_id,
            title: released.title,
            description: released.description,
            time_limit_minutes: released.time_limit_minutes,
            open_at: fmt_utc(released.open_at),
            close_at: fmt_utc(released.close_at),
            show_results_immediately: released.show_results_immediately,
            results_released: released.results_released,
            is_published: released.is_published,
            order_index: released.order_index,
            total_points: released.total_points,
            question_count,
            submission_count,
            grading_period_number: released.grading_period_number,
            component: released.component.clone(),
            tos_id: released.tos_id.clone(),
            created_at: fmt_utc(released.created_at),
            updated_at: fmt_utc(released.updated_at),
        })
    }
}

File: client/lib/data/repositories/assessments/operations/release_results.dart
import 'package:dartz/dartz.dart';
import 'package:likha/core/errors/failures.dart';
import 'package:likha/core/sync/mutation_result.dart';
import 'package:likha/core/sync/sync_queue.dart';
import 'package:likha/core/utils/typedef.dart';
import 'package:likha/data/models/assessments/assessment_model.dart';
import 'package:likha/domain/assessments/entities/assessment.dart';
import 'package:uuid/uuid.dart';
import 'package:likha/data/datasources/local/assessments/assessment_local_datasource.dart';

ResultFuture<MutationResult<Assessment>> releaseResults(
  AssessmentLocalDataSource localDataSource,
  SyncQueue syncQueue, {
  required String assessmentId,
}) async {
  try {
    final (cached, _) = await localDataSource.getCachedAssessmentDetail(assessmentId);
    final now = DateTime.now();

    final optimisticModel = AssessmentModel(
      id: cached.id,
      classId: cached.classId,
      title: cached.title,
      description: cached.description,
      timeLimitMinutes: cached.timeLimitMinutes,
      openAt: cached.openAt,
      closeAt: cached.closeAt,
      showResultsImmediately: cached.showResultsImmediately,
      resultsReleased: true,
      isPublished: cached.isPublished,
      orderIndex: cached.orderIndex,
      totalPoints: cached.totalPoints,
      questionCount: cached.questionCount,
      submissionCount: cached.submissionCount,
      tosId: cached.tosId,
      gradingPeriodNumber: cached.gradingPeriodNumber,
      component: cached.component,
      createdAt: cached.createdAt,
      updatedAt: now,
      syncStatus: SyncStatus.pending,
      cachedAt: now,
    );

    final db = await localDataSource.localDatabase.database;
    await db.transaction((txn) async {
      await localDataSource.releaseResults(assessmentId: assessmentId, txn: txn);
      await syncQueue.enqueue(
        SyncQueueEntry(
          id: const Uuid().v4(),
          entityType: SyncEntityType.assessment,
          operation: SyncOperation.releaseResults,
          payload: {'id': assessmentId},
          status: SyncStatus.pending,
          retryCount: 0,
          maxRetries: 5,
          createdAt: now,
        ),
        txn: txn,
      );
    });

    return Right(MutationResult(entity: optimisticModel, status: SyncStatus.pending));
  } catch (e) {
    return Left(ServerFailure(e.toString()));
  }
}
